src/engine/thread.js

```diff
@@ -1,3 +1,54 @@
+const _stackFrameFreeList = [];
+
+class StackFrame {
+    constructor (warpMode) {
+        // Whether this level of the stack is a loop.
+        this.isLoop = false;
+        // Whether this level is in warp mode.
+        this.warpMode = warpMode;
+        // Reported value from just executed block.
+        this.justReported = null;
+        // Persists reported inputs during async block.
+        this.reported = null;
+        // Name of waiting reporter.
+        this.waitingReporter = null;
+        // Procedure parameters.
+        this.params = null;
+        // A context passed to block implementations.
+        this.executionContext = null;
+    }
+
+    reset () {
+        this.isLoop = false;
+        this.warpMode = false;
+        this.justReported = null;
+        this.reported = null;
+        this.waitingReporter = null;
+        this.params = null;
+        this.executionContext = null;
+        return this;
+    }
+
+    use (warpMode) {
+        this.reset();
+        this.warpMode = warpMode;
+        return this;
+    }
+
+    static create (warpMode) {
+        const stackFrame = _stackFrameFreeList.pop();
+        if (typeof stackFrame !== 'undefined') {
+            stackFrame.warpMode = warpMode;
+            return stackFrame;
+        }
+        return new StackFrame(warpMode);
+    }
+
+    static release (stackFrame) {
+        _stackFrameFreeList.push(stackFrame.reset());
```

Comment:
maybe double check that `stackFrame` exists, since the travis test seems to have failed here calling reset on undefined.